Sources/PubNub/Networking/HTTPFileTask.swift

```diff
@@ -0,0 +1,471 @@
+
+  public struct ProgressUnit {
+    public var currentBytes: Int64
+    public var totalTransmitted: Int64
+    public var totalExpectedToTransmit: Int64
+
+    public init(
+      currentBytes: Int64,
+      totalTransmitted: Int64,
+      totalExpectedToTransmit: Int64
+    ) {
+      self.currentBytes = currentBytes
+      self.totalTransmitted = totalTransmitted
+      self.totalExpectedToTransmit = totalExpectedToTransmit
+    }
+  }
+
+  // Public Responders
+  public var didComplete: ((_ session: URLSessionReplaceable, _ task: URLSessionTask) -> Void)?
+  public var didError: ((_ session: URLSessionReplaceable, _ task: URLSessionTask, _ error: Error) -> Void)?
+  public var didDownload: (
+    (_ session: URLSessionReplaceable, _ task: URLSessionDownloadTask, _ downloadTo: URL) -> Void
+  )?
+  public var didTrasmitData: (
+    (_ session: URLSessionReplaceable, _ task: URLSessionTask, _ update: ProgressUnit) -> Void
+  )?
+
+  // MARK: URLSessionDelegate
+
+  open func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+    if let error = error {
+      didError?(session, task, error)
+      tasksByIdentifier[task.taskIdentifier]?.didError(error)
+    } else {
+      didComplete?(session, task)
+      tasksByIdentifier[task.taskIdentifier]?.didComplete()
+    }
+    // Cleanup Task After Completion
+    tasksByIdentifier.removeValue(forKey: task.taskIdentifier)
+  }
+
+  // MARK: URLSessionDataDelegate
+
+  open func urlSession(
+    _ session: URLSession, task: URLSessionTask,
+    didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64
+  ) {
+    // Progress Updates
+    didTrasmitData?(
+      session,
+      task,
+      ProgressUnit(
+        currentBytes: bytesSent, totalTransmitted: totalBytesSent, totalExpectedToTransmit: totalBytesExpectedToSend
+      )
+    )
+    tasksByIdentifier[task.taskIdentifier]?.updateProgress(
+      bytesWritten: bytesSent,
+      totalBytesWritten: totalBytesSent,
+      totalBytesExpectedToWrite: totalBytesExpectedToSend
+    )
+  }
+
+  open func urlSession(_: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
+    (tasksByIdentifier[dataTask.taskIdentifier] as? HTTPFileUploadTask)?.didReceieve(data: data)
+  }
+
+  // MARK: URLSessionDownloadDelegate
+
+  open func urlSession(
+    _ session: URLSession, downloadTask: URLSessionDownloadTask,
+    didFinishDownloadingTo location: URL
+  ) {
+    guard let fileDownloadTask = tasksByIdentifier[downloadTask.taskIdentifier] as? HTTPFileDownloadTask else {
+      didError?(session, downloadTask, PubNubError(.badRequest))
+      return
+    }
+
+    let temporaryURL = fileDownloadTask.temporaryURL()
+
+    do {
+      // Move file to a temporary location or it's lost at the end of this scope
+      try FileManager.default.moveItem(at: location, to: temporaryURL)
+
+      didDownload?(session, downloadTask, temporaryURL)
```

Comment:
@CraigLn what if we will add same behaviour as `downloadTask(with:completionHandler:)` has (thought, there is completion closure instead of delegate callbacks) where file will be removed from temporary location right after closure call will be completed?

Reply:
I'm not sure.  I only tested how it performs inside the delegate, but can confirm that the URL contents will be deleted at the end of the scope unless moved (even if it's to another location in the `tmp` directory).

Reply:
@CraigLn it is fine. I thought that there can be constraints exactly because it is delegate callbacks.
Than I'm fine with current approach (just wanted some input). 
Thanks!